### ISSUE TITLE
[Feat]: Pull images with background Contexts

### ISSUE THREAD

**Issue by Namnamseo** [project-zot/zot#2505]
### Is your feature request related to a problem? Please describe.

cf. #2416 

When an on-demand sync is triggered, the `ctx` to the call to `SyncImage()` comes from the request `context.Context`.
Thus, when the user gives up, the image sync is also canceled.

Code flow:
https://github.com/project-zot/zot/blob/0d0eae502e438b419a00e8ce4816be76daf8b87a/pkg/api/routes.go#L518
https://github.com/project-zot/zot/blob/0d0eae502e438b419a00e8ce4816be76daf8b87a/pkg/api/routes.go#L1941
https://github.com/project-zot/zot/blob/0d0eae502e438b419a00e8ce4816be76daf8b87a/pkg/extensions/sync/on_demand.go#L76
https://github.com/project-zot/zot/blob/0d0eae502e438b419a00e8ce4816be76daf8b87a/pkg/extensions/sync/service.go#L311

This is unfortunate, I think, in many use cases where Zot is used as a proxy cache.
Even if we couldn't make it on time this time, it isn't that we don't want to have this image cached.

If the client has a timeout limit and a retry loop (like common setups with Kubernetes and containerd do - at least in mine), the current implementation will not be able to make any progress.
This is because, as soon as the client hits the timeout, the sync job fails (because it makes an HTTP request with a canceled Context, I guess), and when the client retries, the sync has to start over again.

It would be better if the pull keeps running in the background, so that the next time there's a request to this image, they could wait for the remaining progress and have any chance of seeing it finish.
Note that the current logic already supports the joining (to wait) to an ongoing pull in another goroutine:
https://github.com/project-zot/zot/blob/0d0eae502e438b419a00e8ce4816be76daf8b87a/pkg/extensions/sync/on_demand.go#L54-L61

### Describe the solution you'd like

Somewhere during the code flow, use a background Context instead of the request Context.
Note that we do not want to completely ignore the request Context; we need to halt the control flow and return when the request is canceled.

A good place could be here:

https://github.com/project-zot/zot/blob/0d0eae502e438b419a00e8ce4816be76daf8b87a/pkg/extensions/sync/on_demand.go#L76-L81

We could give `syncImage` a `context.Background()`, or a larger timeout (15 minutes?).
Then, below, a `select {}` with both `syncResult` and `ctx.Done()` would suffice.

### Describe alternatives you've considered

Since we're spawning a goroutine on every cache miss, this could lead to a large number of background goroutines.
In that prospect, having a full-fledged background process controller (like limiting the number, etc.) could be a long-term goal.

### Additional context

The image size in question is ~10 GiB, and network speed is as low as 10 MiB/s. That would take 1k seconds (= 16m 40s).

**Comment by Namnamseo** [comment]
I didn't go right into writing a PR because I'm not familiar with the code base and am not sure if this is OK.
The leaky goroutine thing feels a bit dangerous so I'd like more opinions on this.

Once there's a resolution, I'd be happy to come up with a PR.

**Comment by peusebiu** [comment]
I agree with your points, you can make this update if you want. I'll review it.

Thank you!

**Comment by andaaron** [comment]
@eusebiu-constantin-petu-dbk does this issue still apply now that https://github.com/project-zot/zot/pull/2903 is merged?

**Comment by apollo13** [comment]
Unstale

**Comment by andaaron** [comment]
@apollo13, @Namnamseo, I was looking at the code and, if the registry configuration has `maxRetries`, those retries are made with the background context. See https://github.com/project-zot/zot/blob/main/pkg/extensions/sync/on_demand.go#L205.

The configuration option at https://github.com/project-zot/zot/blob/main/examples/config-sync.json#L26.

The retry was using a background context since before this issue was opened. Are you saying this retry also gets skipped if the original request context is cancelled?

**Comment by apollo13** [comment]
@andaaron What is the default for `maxRetries`? I don't have it set in my config so I'd assume retries are disabled? Will see about retrying this next week with `maxRetries`. That said the main blocker for me currently seems that it fetches all architectures which for instance for `caddy` amounts to gigabytes of data instead of just a few megs for linux amd64.

**Comment by andaaron** [comment]
> [@andaaron](https://github.com/andaaron) What is the default for `maxRetries`? I don't have it set in my config so I'd assume retries are disabled? Will see about retrying this next week with `maxRetries`. That said the main blocker for me currently seems that it fetches all architectures which for instance for `caddy` amounts to gigabytes of data instead of just a few megs for linux amd64.

It's disabled by default.

**Comment by lfrancke** [comment]
I just created a PR for this: https://github.com/project-zot/zot/pull/3537

**Comment by andaaron** [comment]
I tweaked the timeouts in https://github.com/project-zot/zot/pull/3574, it is merged to main, but not yet released.

**Comment by lfrancke** [comment]
I think this can be closed now. I don't have the permission to do so.
